### PR TITLE
Support pad

### PR DESCRIPTION
**Pull request by juliuskunze**
~~Simultaneous negative and interior padding is not yet supported.~~

### CODE REVIEW

File: fastar/rules.py [hunk 1]
```diff
@@ -250,3 +248,36 @@ def dot_general_backward_rule(outbox, lhs, rhs, dimension_numbers, precision):
...
+  dilation = np.array(interior) + 1
+  inclip = lambda indices: np.clip(indices, 0, operand.shape)
+  next_instart = inclip(lax.lax._ceil_divide(out_start - lo, dilation))
+  next_outstart = next_instart * dilation + lo
+  _start = next_outstart - out_start
+  _stop = _start + jnp.array(incount.shape) * dilation
+  out_index = [slice(start, stop, step)
+               for start, stop, step in zip(_start, _stop, dilation)]
+  out_slice = jnp.full(out_shape, padding_value)
+  out_slice = ops.index_update(out_slice, out_index, inslice)
```
Comment (j-towns) [inline fastar/rules.py]: I'm just wondering whether we can express the creation of out_slice in terms of lax.pad, I'm having a go at it now.

Comment (juliuskunze) [reply fastar/rules.py]: Totes, good idea!

Comment (j-towns) [reply fastar/rules.py]: @JuliusKunze do you wanna do it actually? I spent 10 mins or so trying to do it last night but didn't get very far.

File: fastar/rules.py [hunk 2]
```diff
@@ -250,3 +248,36 @@ def dot_general_backward_rule(outbox, lhs, rhs, dimension_numbers, precision):
   return [lhs_start, rhs_start], [lhs_incount, rhs_incount]
 
 def_standard_op(lax.dot_general_p, dot_general_backward_rule)
+
+def pad_backward_rule(outbox, operand, padding_value, padding_config):
+  out_start, out_shape = outbox
+  lo, _, interior = unzip3(padding_config)
```
Comment (j-towns) [inline fastar/rules.py:254]: One slight optimization we could do here is to catch when the outbox only contains padded values (i.e. it has no dependence on the input data) and return `None` for the operand instart and count. That case would also have a simpler update rule.

File: fastar/rules.py (hunk 1 above)
Comment (j-towns) [reply fastar/rules.py]: Generally I think we should attempt to avoid anything which bottoms out on a gather, scatter or reshape, because they're known to be slow to compile, and therefore slow in op-by-op.

Comment (juliuskunze) [reply fastar/rules.py]: Sure, will do

Comment (juliuskunze) [reply fastar/rules.py]: Done!

File: fastar/rules.py (hunk 2 above)
Comment (juliuskunze) [reply fastar/rules.py:254]: Done!

**Comment by j-towns** [comment]
Brilliant work Julius!
